alibi/explainers/ale.py

```diff
@@ -0,0 +1,66 @@
+import numpy as np
+from typing import Callable
+
+
+def get_quantiles(values: np.ndarray, num_points: int = 11) -> np.ndarray:
+    """
+    Calculate quantiles of values in an array.
+
+    Parameters
+    ----------
+    values
+        Array of values
+    num_points
+        Number of quantiles to calculate
+
+    Returns
+    -------
+    Array of quantiles of the input values
+
+    """
+    percentiles = np.linspace(0, 100, num=num_points)
+    quantiles = np.percentile(values, percentiles, axis=0)
+    return quantiles
+
+
+def first_ale_num(predict: Callable, X: np.ndarray, feature: int, num_points: int = 11) -> np.ndarray:
+    """
+    Calculate the first order ALE for a numerical feature.
+
+    Parameters
+    ----------
+    predict
+        Model prediction function
+    X
+        Dataset to calculate the ALE on
+    feature
+        Index of the numerical feature for which to calculate ALE
+    num_points
+        Number of quantiles to calculate for the feature
+
+    Returns
+    -------
+    ALE values for the feature
+
+    """
+    # TODO the following onlt works for regression
+    q = get_quantiles(X[:, feature], num_points=num_points)
+    ale = np.zeros(len(q) - 1)
+
+    for i in range(len(q) - 1):
+        # TODO this misses the data point equal to the upper quantile
+        subset = X[(q[i] <= X[:, feature]) & (X[:, feature] < q[i + 1])]
```

Comment:
Just a suggestion (which takes care of the TODO)

```
    indices = np.digitize(X[:, feature], q)
    for i in range(1, num_points):
        if i == num_points - 1:
            subset = X[indices >= i]
        else:
            subset = X[indices == i]
```

Reply:
This wouldn't quite work, first the order would be reversed (using side="right" would be closer to the original code. More importantly, this will still not pick up the last point (e.g. x[indices==num_points]. I think the point equal to the last quantile should be merged with the previous quantile, although it's probably not a big deal if it's not used at all.

Reply:
First - you're right, the bottom snipped should use `side="right"`, sorry about that, but I think the top snippet works.

If we start with
```
import numpy as np

np.random.seed(42)
X = np.random.randn(5, 2)
feature = 0
values = X[:, feature]
num_points = 3

percentiles = np.linspace(0, 100, num=num_points)
q = np.percentile(values, percentiles, axis=0)
```

then

```
for i in range(len(q) - 1):
    subset = X[(q[i] <= X[:, feature]) & (X[:, feature] < q[i + 1])]
    print(subset)
```
and
```
indices = np.searchsorted(q, X[:, feature], side="right")
for i in range(1, num_points):
    subset = X[indices == i]
    print(subset)
```
produce the same output.

Both miss the last point, so to capture it, we can change it to
```
indices = np.searchsorted(q, X[:, feature], side="right")
for i in range(1, num_points):
    if i == num_points - 1:
        subset = X[indices >= i]
    else:
        subset = X[indices == i]
    print(subset)
```
then we do capture the point equal to the uppermost quantile (here, `[ 1.57921282  0.76743473]`).